File: dynamic-programming/bottom-up/stock-trading/SingleStockTrader.java

public class SingleStockTrader {
    public int maxProfit(int[] prices) {
        if (prices == null || prices.length == 0) {
            return 0;
        }
        int sold1Prft = 0;
        int hold1Prft = Integer.MIN_VALUE;
        for (int price : prices) {
            sold1Prft = Math.max(sold1Prft, hold1Prft + price);
            hold1Prft = Math.max(hold1Prft, -price);
        }
        return sold1Prft;
    }
}

File: dynamic-programming/bottom-up/stock-trading/SingleStockTrader.cpp
class SingleStockTrader {
public:
    int maxProfit(vector<int>& prices) {
        if (prices.size() == 0) {
            return 0;
        }
        int sold1Prft = 0;
        int hold1Prft = INT_MIN;
        for (int curPrice : prices) {
            sold1Prft = max(sold1Prft, hold1Prft + curPrice);
            hold1Prft = max(hold1Prft, -curPrice);
        }
        return sold1Prft;
    }
};
